dhm: Fix bias in private key generation

ronald-cron-arm opened the pull request:
## Description
Fix #4612

## Status
READY

ronald-cron-arm commented:
@gilles-peskine-arm @mpg I believe the commit in this PR addresses in a minimalist way #4612. Do we want to do more like fixing the equivalent bias in blinding values in dhm.c ? in ecp.c ?

gilles-peskine-arm commented:
> @gilles-peskine-arm @mpg I believe the commit in this PR addresses in a minimalist way #4612. Do we want to do more like fixing the equivalent bias in blinding values in dhm.c ? in ecp.c ?

Ideally yes, if it can be done without too much trouble.

gilles-peskine-arm commented on library/dhm.c:
```diff
@@ -198,19 +198,18 @@ int mbedtls_dhm_make_params( mbedtls_dhm_context *ctx, int x_size,
             return( MBEDTLS_ERR_DHM_MAKE_PARAMS_FAILED );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_fill_random( &ctx->X, x_size, f_rng, p_rng ) );
+        MBEDTLS_MPI_CHK( mbedtls_mpi_shift_r( &ctx->X, 8 * x_size - n_bits ) );
```
This is correct when `x_size` is the size of `P`. But when `x_size` is smaller, it's reducing the range. (`x_size` larger than the size of `P` doesn't make sense, so it's ok if the behavior changes in that case, although in an LTS branch we should be conservative even in error cases.)

I think this would be a correct strategy:
```
if( 8 * x_size < n_bits )
    MBEDTLS_MPI_CHK( mbedtls_mpi_fill_random( &ctx->X, x_size, f_rng, p_rng ) );
else
{
    // the loop, with 8 * n_bits instead of x_size
}
```

gilles-peskine-arm commented on library/dhm.c:
```diff
@@ -316,19 +316,18 @@ int mbedtls_dhm_make_public( mbedtls_dhm_context *ctx, int x_size,
             return( MBEDTLS_ERR_DHM_MAKE_PUBLIC_FAILED );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_fill_random( &ctx->X, x_size, f_rng, p_rng ) );
+        MBEDTLS_MPI_CHK( mbedtls_mpi_shift_r( &ctx->X, 8 * x_size - n_bits ) );
```
Even in an LTS branch, I think unifying the code of `mbedtls_dhm_make_params` and `mbedtls_dhm_make_public`, which differs only by the choice of error code, would be better.

gilles-peskine-arm reviewed (commented):
* I am satisfied that this is a complete backport (with minimal necessary adaptations) of the part of #4276 that concerns Diffie-Hellman, except for the change to use the new function `mbedtls_mpi_random` in `dhm_random_below` and the changelog entry.
* I am satisfied that the commit "dhm: Fix bias in private key generation and blinding" is correct and does the same job as the part of #4276 concerning `dhm.c` that was not backported.
* I am satisfied that the changes are not too disruptive for a LTS branch. I have slight concerns about "DHM refactoring: unify mbedtls_dhm_make_{params,public}" changing error codes in corner cases, but I can't identify one in a reasonable amount of time, so I think they would be too exotic to matter even by LTS standards.

So all that's missing for my approval is a changelog entry.

ronald-cron-arm commented:
Thanks for the feedback @gilles-peskine-arm. For ECP, I propose to just fix the blinding values generation in ecp_randomize_jac/mxz(). I've added a commit for that. Otherwise I've added the change log entry as in #4276.

mpg reviewed (approved):
Looks good to me. I've verified that most of the commits are from the original PR (thanks for including details about the changes in the commit message!), and that all relevant commits have been backported. I've then review the two new commits independently, and I'm happy with the result.
